pkg/apis/configuration/v1/configsource.go

```diff
@@ -7,13 +7,33 @@ type ConfigSource struct {
 	SecretValue SecretValueFromSource `json:"secretKeyRef,omitempty"`
 }
 
+// ConfigPatch is a JSON patch to add values from secrets in the same namespace with KongPlugin
+// to the generated configuration of plugin in Kong.
+// +kubebuilder:object:generate=true
+type ConfigPatch struct {
+	// Path is the JSON path to add the patch.
+	Path string `json:"path"`
+	// ValueFrom is the reference to a key of a secret where the patched value comes from.
+	ValueFrom ConfigSource `json:"valueFrom"`
+}
+
 // NamespacedConfigSource is a wrapper around NamespacedSecretValueFromSource.
 // +kubebuilder:object:generate=true
 type NamespacedConfigSource struct {
 	// Specifies a name, a namespace, and a key of a secret to refer to.
 	SecretValue NamespacedSecretValueFromSource `json:"secretKeyRef,omitempty"`
 }
 
+// NamespacedConfigPatch is a JSON patch to add values from secrets to KongClusterPlugin
+// to the generated configuration of plugin in Kong.
+// +kubebuilder:object:generate=true
+type NamespacedConfigPatch struct {
+	// Path is the JSON path to add the patch.
+	Path string `json:"path"`
+	// ValueFrom is the reference to a key of a secret where the patched value comes from.
+	ValueFrom NamespacedConfigSource `json:"valueFrom"`
```

Comment:
Why not create a single type for `ValueFrom` with an optional `Namespace` field? That would allow having a single `ConfigPatch` type. I think it would be more user-friendly approach. 🤔 

When it comes to validation, we could make `Namespace` required when used in `KongClusterPlugin` with the use of CEL rules.

Reply:
This changes the existing `ConfigFrom` because it uses `NamespacedConfigSource`. If we think It is OK to do such a change, we could unify the definition of `ConfigSource`. (Considering that we should deprecate `ConfigFrom` when this is implemented)

Reply:
I see. Probably `NamespacedConfigSource` made more sense when we could not use CEL. 

I wouldn't like to block this change, but we can discuss the possibility of transitioning to having a common structure for namespaced/non-namespaced sources. As I understand, removing Go type for `NamespacedSecretValueFromSource` and using `SecretValueFromSource` with the added optional `Namespace` field would be possible without breaking changes in CRDs as this would be:
- adding an optional `namespace` field to `KongPlugin`'s `configFrom.secretKeyRef` (adding fields is not breaking)
- making `Namespace` in `KongClusterPlugin`'s `configFrom.secretKeyRef` optional (and requiring it via CEL) which also would not be breaking as I understand.

Reply:
Let's discuss this in a new issue.